file.go

```diff
@@ -0,0 +1,259 @@
+		return "", nil, fmt.Errorf("failure to seek to the metadata root: %v", err)
+	}
+
+	read := func(data interface{}) bool {
+		err = binary.Read(r, binary.LittleEndian, data)
+		return err == nil
+	}
+	readStr := func(n int, data *string) bool {
+		buf := make([]byte, n)
+		err = binary.Read(r, binary.LittleEndian, buf)
+		if err == nil {
+			*data = cstring(buf)
+		}
+		return err == nil
+	}
+
+	// the Metadata header is defined in §II.24.2.1.
+	var hdr struct {
+		Signature    uint32
+		MajorVersion uint16
+		MinorVersion uint16
+		Reserved     uint32
+		Version      string
+		Flags        uint16
+	}
+	if !read(&hdr.Signature) {
+		return "", nil, fmt.Errorf("failure to read the metadata header signature: %v", err)
+	}
+	// magic signature from II.24.2.1
+	const signature = 0x424A5342
+	if hdr.Signature != signature {
+		return "", nil, fmt.Errorf("metadata header signature (%#X) must be (%#X)", hdr.Signature, signature)
+	}
+	var streamsCount uint16
+	var cstringLength uint32
+	if !read(&hdr.MajorVersion) ||
+		!read(&hdr.MinorVersion) ||
+		!read(&hdr.Reserved) ||
+		!read(&cstringLength) ||
+		!readStr(int(cstringLength), &hdr.Version) ||
+		!read(&hdr.Flags) ||
+		!read(&streamsCount) {
+		return "", nil, fmt.Errorf("failure to read the metadata header: %v", err)
+	}
+
+	readStreamNameStr := func(data *string) bool {
+		// the name of the stream as null-terminated variable length array
+		// of ASCII characters, padded to the next 4-byte boundary
+		// with \x00 characters. The name is limited to 32 characters.
+		const nameMaxLength = 32
+		const namePadding = 4
+		var nameBuf [nameMaxLength]byte
+		// Read in chunks of 4 bytes, accumulating the string
+		// into nameBuf until the first \x00 character is found.
+		var found bool
+		for j := 0; j < nameMaxLength; j += namePadding {
+			if !read(nameBuf[j : j+namePadding]) {
+				break
+			}
+			idx := bytes.IndexByte(nameBuf[:j+namePadding], 0)
+			if idx != -1 {
+				found = true
+				*data = string(nameBuf[:idx])
+				break
+			}
+		}
+		if !found {
+			err = errors.New("name not found")
+		}
+		return err == nil
+	}
+
+	// parse stream headers.
+	heaps := make([]*Heap, streamsCount)
+	for i := 0; i < int(streamsCount); i++ {
+		// the stream header is defined in §II.24.2.2.
+		var s struct {
+			Offset uint32
+			Size   uint32
+			Name   string
+		}
+		if !read(&s.Offset) ||
+			!read(&s.Size) ||
+			!readStreamNameStr(&s.Name) {
+			return "", nil, fmt.Errorf("failure to read the stream header (%d): %v", i, err)
+		}
+		sr := io.NewSectionReader(ds, rootOffset+int64(s.Offset), int64(s.Size))
+		heaps[i] = &Heap{
+			sr:       sr,
+			ReaderAt: sr,
+			name:     s.Name,
+			Size:     s.Size,
+		}
+	}
+	return hdr.Version, heaps, nil
+}
+
+// sectionByRVA returns the section which contains rva.
+func sectionByRVA(pefile *pe.File, rva uint32) *pe.Section {
+	var ds *pe.Section
+	for _, s := range pefile.Sections {
+		if s.VirtualAddress <= rva && rva < s.VirtualAddress+s.VirtualSize {
```

Comment:
Bit of a meta question: how concerned do we want to be about invalid data here? Essentially looking at this code and thinking about what happens when `s.VirtualAddress+s.VirtualSize` overflows. 

For our current purposes that doesn't matter too much because we're dealing with known good data. For a more general purpose usage though not sure. Frankly unsure how the .NET equivalent handels this.

Reply:
Another invalid-per-spec scenario that comes to mind is: what if there are multiple `#Strings` heaps? Currently the code picks the last one it finds, although we *could* detect it fairly easily and fail, if we wanted.

The current scenario for the library is to generate code that matches up with the winmd, so I suppose detecting invalid data would potentially help diagnose issues if the original winmd generation went horribly wrong.

Reply:
> Bit of a meta question: how concerned do we want to be about invalid data here? Essentially looking at this code and thinking about what happens when s.VirtualAddress+s.VirtualSize overflows.

Checking if `s.VirtualAddress+s.VirtualSize` will overflow cost nothing and could help debug invalid winmd files, I'll do that. Well, it can't happen on x64, only on x86, as both types are `uint32`, but anyway.

> Frankly unsure how the .NET equivalent handels this.

.NET is not checking for overflows:

https://github.com/dotnet/runtime/blob/3c389db900f949122f4ad51f02cadc709bf12815/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeaders.cs#L349-L361

Reply:
> Another invalid-per-spec scenario that comes to mind is: what if there are multiple `#Strings` heaps? Currently the code picks the last one it finds, although we _could_ detect it fairly easily and fail, if we wanted.
> 
> The current scenario for the library is to generate code that matches up with the winmd, so I suppose detecting invalid data would potentially help diagnose issues if the original winmd generation went horribly wrong.

Good catch! Will do that too.